packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -18,42 +18,310 @@ interface Window {
   };
 }
 
-type Registry = {
-  [url: string]: Module
-};
+type ResolveCallback = (...args: Array<{}>) => void;
+type ErrorCallback = (error: Error) => void;
+type NormalizedUrl = string&{_normalized: never};
+
+(function() {
+
+/**
+ * Describes the state machine for loading a single module.
+ *
+ * At runtime, these are integers that are inlined in their use sites.
+ */
+const enum StateEnum {
+  /**
+   * The initial state.
+   */
+  Initialized,
+
+  /**
+   * Comes after Initialized. We have begun loading the module over the network.
+   * Toplevel scripts skip this state entirely.
+   */
+  Loading,
+
+  /**
+   * Comes after Loading. The module's <script> tag has loaded. If there was a
+   * define() call for this module, it has run by now. We have started loading
+   * the module's dependencies, but we can't execute them until some other,
+   * earlier scripts have executed.
+   */
+  WaitingOnEarlierScripts,
+
+  /**
+   * Comes after WaitingOnEarlierScripts. All earlier scripts are now executed,
+   * and we can now execute our dependencies in order. Once that's done, we can
+   * execute this module too.
+   */
+  WaitingOnDeps,
+
+  /**
+   * The successful terminal state. Comes after WaitingOnDeps. All of the
+   * module's dependencies have loaded and executed, and the module's body, if
+   * any, has executed, and there were no errors.
+   */
+  Executed,
+
+  /**
+   * The unsuccessful terminal state. Something went wrong, either leading up
+   * to, or during execution of the module.
+   */
+  Failed,
+}
 
-interface Module {
+interface Initialized {
+  state: StateEnum.Initialized;
+}
+interface Loading {
+  state: StateEnum.Loading;
+}
+interface BaseWaiting {
+  /**
+   * The dependencies of this module, in order.
+   */
+  deps: Module[];
+  /**
+   * Args that we will pass into to moduleBody.
+   */
+  args: Array<{}>;
+  /**
+   * The body of the module, which is executed after its dependencies are
+   * loaded.
+   *
+   * In AMD/Commonjs terminology, this is the factory function.
+   */
+  moduleBody: undefined|Function;
```

Comment:
I'd rather name it as the factory then

Reply:
We're using ES standard terminology in other places (e.g. we call them dependency specifiers, not ids), which I think is desirable.

I also think it's really weird to refer to a function that's only called once and doesn't return anything as a factory.